# LEET CODE 1. Two Sum

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

In [2]:
nums = [2,7,11,15]
target = 9

### Brute Force Solution

In [3]:
def twoSum(nums, target):

    for i in range(0,len(nums)):
        for j in range(i+1, len(nums)):

            y = nums[i] + nums[j]

            if y == target:
                result = [i,j]
                return result

result = twoSum(nums, target)
print(result)

[0, 1]


# Hashmap

A HashMap is a data structure that provides an efficient way to store and retrieve **key-value pairs**.

1.	Efficiency:
	•	HashMaps offer efficient data storage and retrieval, which is critical in many data science tasks like counting frequencies, aggregating data, and performing quick lookups.
2.	Common Interview Problems:
	•	Many coding interview questions in data science can be solved more efficiently with HashMaps. Examples include:
	•	Counting occurrences (e.g., word count).
	•	Finding duplicates.
	•	Two-sum problem (checking if two numbers add up to a target).
	•	Grouping and aggregating data based on keys.

In [4]:
def twoSum_hashmap(nums, target):
    
    map = {} # Initialize and empty Dictionary
    
    # Iterate through each element in the nums list
    for i in range(len(nums)):

        complement = target - nums[i] # the value we need to find in the list

        # checks if the value of complement exists as a key in the dictionary map.
        if complement in map:
            return (   map[complement], i  )
        
        #Update the map dictionary
        map[nums[i]] = i # add the current number (nums[i]) as the key and index (i) as the value

    return [] 

twoSum_hashmap(nums, target)

(0, 1)

### A quick note on dictionaries in Python

A dictionary in Python is a built-in data structure that stores data as key-value pairs. Key concepts:

1.	Key-Value Pairs:
	•	Each element in a dictionary is stored as a pair consisting of a unique key and an associated value.
	•	Example: {"apple": 3, "banana": 5} — Here, "apple" and "banana" are keys, and 3 and 5 are their corresponding values.

2.	Keys:
	•	Must be unique and immutable (e.g., strings, numbers, or tuples).
	•	Used to access the corresponding value efficiently.
	•	In the two-sum problem, keys represent numbers encountered in the list (nums).

3.	Values:
	•	Can be of any data type and are associated with keys.
	•	In the two-sum problem, values represent indices of the numbers in the list.

4.	Initialization and Access:
	•	A dictionary is created using curly braces {}.
	•	You can add or access elements using square brackets [].
	•	Example: map = {} initializes an empty dictionary. map[2] = 0 adds a key 2 with value 0.

5.	Lookup:
	•	if key in dictionary: checks if a key is present in the dictionary.
    To check if a value exists in a dictionary in Python, we have to use "x in map.values()"


### 349. Intersection of Two Arrays
Given two integer arrays nums1 and nums2, return an array of their  intersection. 
Each element in the result must be unique and you may return the result in any order.


Examples

Input: nums1 = [1,2,2,1], nums2 = [2,2].      |Output: [2]

Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]  | Output: [9,4]

In [1]:
nums1 = [4,9,5, 9]
nums2 = [9,4,9,8,4]

class Solutions:
    def intersection(self,nums1, nums2):
        """
        Find the intersection of two integer arrays and return the result as a list.
        Each element in the result must be unique.
        Complexity:
            - Time: O(n+m)
            - Space O(n)
        """
        # Initialize 
        is_available = {} # Dictionary to record elements from nums1 
        result = []     # List to store the unique intersection elements
        
        # record each element from nums1 in the dictionary ...
        for x in nums1:
            is_available[x] = 1   # mark value as available

        # Iterate over nums2 to find common elements 
        for x in nums2:

            # If element in nums 2 as is available:
            if x in is_available and is_available[x] == 1:
                result.append(x)        # Add to results
                is_available[x] = 0     # Mark as unaveilable

        return result
    # Aditional explanation. in a Python dictionary, the keys are always unique. 
    # Each key in a dictionary maps to a single value. If you attempt to assign 
    # a value to an existing key, the old value associated with that key will be
    #  overwritten by the new value. Therefore, each value in nums1 will appear 
    # only once is is_avaliable. That will guarantee uniqueness of the results

    # Alternativelly, we could've used built in set in Python

    def intersection_set(self, nums1, nums2):

        set1 = set(nums1)
        set2 = set(nums2)
        return set1 & set2   



### 350:  Intersection of Two Arrays II

Given two integer arrays nums1 and nums2, return an array of their intersection. 
Each element in the result must appear as many times as it shows in both arrays 
and you may return the result in any order.

 

Example 1:

Input: nums1 = [1,2,2,1], nums2 = [2,2]
Output: [2,2]
Example 2:

Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
Output: [4,9]
Explanation: [9,4] is also accepted.

In [5]:
# This solution has a time complexity of O(n * m), since I have a nested loop
# the solution bellow is both easier and more efficient

class Solution:
    def intersect(self, nums1, nums2) -> int:
        """
        This function finds the intersection of two lists, nums1 and nums2, 
        and return the elements that are common to both lists.
        Done by myself =)!!
        """     
        
        # Initialize empty list to store the intersection values
        results = []
        
        
        # Create a dictionary (nums2_dict) from nums2 using enumerate
        # The dictionary maps the index of each element in nums2 to the element itself
        # Example: nums2 = [4, 9, 5] -> nums2_dict = {0: 4, 1: 9, 2: 5}

        nums2_dict = dict(enumerate(nums2))

        # Iterate over each element in nums1
        for i in nums1:
            
            # Initialize/update variable to store the key (index) of the found value
            key_found = None
            
            # Iterate through each key-value pair in the nums2_dict dictionary
            for key, value in nums2_dict.items():

                # Check if the current value from nums2_dict matches the current element from nums1 (i)
                if value == i:

                    # If a match is found, store the key of the found value in key_found
                    key_found = key
                    
                     # Append the found value to the results list
                    results.append(value)

                    # Delete the found key-value pair from nums2_dict to avoid duplicate matches
                    del nums2_dict[key_found]

                    # Break the inner loop since the value has been found and handled
                    break

            # Return the list of intersected values    
        return results

nums1 = [4,9,5,9, 10]
nums2 = [9,4,9,8,4,10]            

a = Solution().intersect(nums1, nums2)
print(a)


[4, 9, 9, 10]


In [ ]:
class Solution:
    def intersect(self, nums1, nums2):
        """
        This function finds the intersection of two lists, nums1 and nums2,
        and returns the elements that are common to both lists, including duplicates.
        """

        # 1) Count the ocurrency frequency of nums2 elements
        nums2_counts = {}
        for num in nums2:
            if num in nums2_counts:
                nums2_counts[num] = nums2_counts[num] +1
            else:
                nums2_counts[num] = 1

        # Iterate over nums1 and find common elements
        results = []
        for num in nums1:
            # Check if the element is in nums2_counts and has a positive count
            if nums2_counts.get(num,0) > 0:
                results.append(num) # Add the element to the results list
                nums2_counts[i] = nums2_counts[i] -1 # Reduce the count in nums2_counts

        return results

Note: The .get() method is a built-in function that allows you to retrieve the 
value for a given key from a dictionary. It provides a safe way to access 
dictionary values, offering a default value if the key does not exist, which 
helps prevent common errors in your code.

Syntax:

`value = my_dict.get(key, default_value)`

* key: the key you want to look up in the dictionary
* default_value (optional): the value to return if the key is not found (None is
the default)

nums2_dict.values

# 88. Merge Sorted Array

You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.

Merge nums1 and nums2 into a single array sorted in non-decreasing order.

The final sorted array should not be returned by the function, but instead be stored inside the array nums1. To accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that should be merged, and the last n elements are set to 0 and should be ignored. nums2 has a length of n.

 

Example 1:

Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n = 3
Output: [1,2,2,3,5,6]
Explanation: The arrays we are merging are [1,2,3] and [2,5,6].
The result of the merge is [1,2,2,3,5,6] with the underlined elements coming from nums1.

#### Brute force (aka real world) solution

In [6]:
nums1 = [1,2,3,0,0,0]
m = 3
nums2 = [2,5,6]
n = 3



for i in range(m, m+n):
    nums1[i] = nums2[m-i]

nums1.sort()

print(nums1)


[1, 2, 2, 3, 5, 6]


#### Three Pointers (start from the end)
This is the most efficient solution

In [7]:
nums1 = [1,2,3,0,0,0] 
nums2 = [2,5,6]       
m = 3
n = 3

nums1 = [2,0]
m = 1
nums2 = [1]
n = 1


# Initialize read pointers for nums1 and nums2, starting from their end positions.

p1 = m-1
p2 = n-1

# and writer point at the very end of num1
p = m+n -1

# Continue the loop while there are elements to compare in nums2 (p2 >= 0).
# The loop also checks (p >= 0) to ensure we don't go out of bounds in nums1.
while (p >= 0) & (p2 >=0) :


    if (nums2[p2] > nums1[p1]) | (p1 < 0):
        """
        If the current element in nums2 is greater than the current element in nums1,
        OR if all elements from nums1 have been placed and p1 is out of bounds (p1 < 0),
         then place the element from nums2 into the current position of nums1.
        This handles the case where we have finished processing nums1 but still have
         elements left in nums2.
        """


        nums1[p] = nums2[p2]  # Place nums2's element in nums1 at location p.
        p2 = p2 - 1           # Move pointer p2 to the left in nums2.
    
    else:
        nums1[p] = nums1[p1] # Place nums1's element p1 in position p in nums1.
        p1 = p1 -1           # Move pointer p1 to the left in nums1.
    
     # Move write pointer `p` to the left to fill the next position.
    p = p-1 

print(nums1)


[1, 2]


# 238. Product of Array Except Self
% Topics: Array, Prefix Sum

Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.


Example 1:

Input: nums = [1,2,3,4]
Output: [24,12,8,6]
Example 2:

Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]

 

In [11]:

# A) Brute Force
nums = [-1,1,0,-3,3]

result = []

for i in range(len(nums)):
    product = 1
    for j in range(len(nums)):
        if i != j:
            product = product * nums[j]
    
    result.append(product)

[0, 0, 9, 0, 0]


For every given index, i, we will make use of the product of all the numbers to 
the left of it and multiply it by the product of all the numbers to the right. 
This will give us the product of all the numbers except the one at the given 
index i.
<img src="images/238_left.png" alt="Left" width="450"/>
<img src="images/238_right.png" alt="right" width="450"/>



1. We initialize lists left and right such that we can access and assign new 
values throughout the loop. THis is a usefull alternative than the usual append

2. Notice the second loop:  `for i in range(n - 2, -1, -1)`. This means that we 
go from n-2 to 0 taking steps of size 1, but in *reverse order*. 

In [39]:
# Left and Right Products
nums = [4, 5, 1, 8, 2, 10, 6]
n = len(nums)

left = [1] * n
right = [1] * n

for i in range(1, n): 
    left[i] = left[i - 1] * nums[i - 1]
    #print(f"i:{i}, left:{left}, nums:{nums} ")

for i in range(n - 2, -1, -1):
    right[i] = right[i + 1] * nums[i + 1]
    #print(f"i:{i}, right:{right}, nums:{nums} ")

answer = [left[i] * right[i] for i in range(n)]

print(f"answer: {answer}")


i:1, left:[1, 4, 1, 1, 1, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:2, left:[1, 4, 20, 1, 1, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:3, left:[1, 4, 20, 20, 1, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:4, left:[1, 4, 20, 20, 160, 1, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:5, left:[1, 4, 20, 20, 160, 320, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:6, left:[1, 4, 20, 20, 160, 320, 3200], nums:[4, 5, 1, 8, 2, 10, 6] 
i:5, right:[1, 1, 1, 1, 1, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:4, right:[1, 1, 1, 1, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:3, right:[1, 1, 1, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:2, right:[1, 1, 960, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:1, right:[1, 960, 960, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
i:0, right:[4800, 960, 960, 120, 60, 6, 1], nums:[4, 5, 1, 8, 2, 10, 6] 
answer: [4800, 3840, 19200, 2400, 9600, 1920, 3200]


# 7. Reverse Integer

Theme: math.
Given a signed 32-bit integer x, return x with its digits reversed. If reversing 
x causes the value to go outside the signed 32-bit integer range [-231, 231 - 1],
then return 0.

* Example 1: Input: x = 123. | Output: 321

* Example 2:  Input: x = -123 | Output: -321

* Example 3: Input: x = 120 | Output: 21

In [102]:
class Solution:
    def reverse(self, x:int) -> int:
        if x > 0:
            x_str = str(x)
            x_str = x_str[::-1]
            x_rvd = int(x_str)
        else:
            x_str = str(x*(-1))
            x_str = x_str[::-1]
            x_rvd = int(x_str)*(-1)

        return x_rvd if x_rvd > (-2**31) and x_rvd < 2 ** 31 else 0

x = -432

Solution().reverse(x)

-234

### Slicing in Python

Notice that we use the slice [::-1] in the solution above. Here is a quick 
summary of slicing in Python:

Slicing allows you to access a subset of elements from sequences such as strings, 
lists, tuples, and other objects that support the sequence protocol.

#### Basic Syntax
`sequence[start:stop:step]``
	• start: The index where the slice starts (inclusive). If omitted, the 
	slice starts from the beginning.
	• stop: The index where the slice ends (exclusive). If omitted, the slice 
	goes up to the end of the sequence.
	• step: The step size or stride. If omitted, it defaults to 1. A negative
	 step reverses the direction.

`my_list = [0, 1, 2, 3, 4, 5]`

1. Basic Slicing	
	* `slice1 = my_list[1:4]`  
	Output: [1, 2, 3 ]

2. Omitting start or stop
	* `slice2 = my_list[:3]`  
	 Output: [0, 1, 2]
	* `slice3 = my_list[3:]`  
	 Output: [3, 4, 5]

3. Using steps to control the stride
	* Every second elevent
	`slice5 = my_list[::2]`
	Output: [0, 2, 4]

4. Steps = -1 to reverse the sequence
	* `reversed_list = my_list[::-1]`  
	outputs [5, 4, 3, 2, 1, 0]

####  9. Palindrome Number
Given an integer x, return true if x is a 
palindrome, and false otherwise.

Input: x = 121 | Output: true
Explanation: 121 reads as 121 from left to right and from right to left.


Input: x = -121 | Output: false
Explanation: From left to right, it reads -121. From right to left, it becomes 121-. Therefore it is not a palindrome.

In [152]:
class Solution:
    def isPalindrome(self, x:int) -> bool:
        x_str = str(x)
        x_rvd = x_str[::-1]
        x_rvd

        if x_rvd == x_str:
            return True
        else:
            return False




Not a palindrome!


In [ ]:
def isPossible(self, A):
    """
    Determines if it's possible to split the sorted integer array 'A' into one or more subsequences
    such that each subsequence is a consecutive increasing sequence of length at least 3.

    Parameters:
    - A: List[int] - A sorted list of integers.

    Returns:
    - bool: True if possible to split as per the conditions, False otherwise.
    """
    import collections  # Import the 'collections' module to use 'Counter'.

    # 'left' counts the number of times each number appears in 'A'.
    # It represents the numbers that are still available to be placed in subsequences.
    left = collections.Counter(A)

    # 'end' counts the number of subsequences ending with a specific number.
    # It helps track subsequences that can be extended.
    end = collections.Counter()

    # Iterate over each number 'i' in the array 'A'.
    for i in A:
        # If 'i' has already been used up in subsequences, skip it.
        if not left[i]:
            continue

        # Use one occurrence of 'i'.
        left[i] -= 1

        # Check if there's a subsequence ending with 'i - 1' that can be extended.
        if end[i - 1] > 0:
            # Decrement the count of subsequences ending with 'i - 1'.
            end[i - 1] -= 1
            # Increment the count of subsequences now ending with 'i'.
            end[i] += 1
        # If no such subsequence exists, try to create a new subsequence [i, i+1, i+2].
        elif left[i + 1] > 0 and left[i + 2] > 0:
            # Use one occurrence of 'i + 1'.
            left[i + 1] -= 1
            # Use one occurrence of 'i + 2'.
            left[i + 2] -= 1
            # Increment the count of subsequences ending with 'i + 2'.
            end[i + 2] += 1
        else:
            # Cannot place 'i' in any subsequence as per the rules.
            return False  # Early exit since the conditions cannot be met.

    # All numbers have been placed successfully in valid subsequences.
    return True

### 659. Split Array into Consecutive Subsequences
https://leetcode.com/problems/split-array-into-consecutive-subsequences/description/?envType=company&envId=google&favoriteSlug=google-all&difficulty=EASY%2CMEDIUM&role=data-scientist-data-engineer

Medium / hard

You are given an integer array nums that is **sorted in non-decreasing order.**

Determine if it is possible to split nums into **one or more subsequences** such
that both of the following conditions are true:

1. Each subsequence is a consecutive increasing sequence (i.e. each integer is 
exactly one more than the previous integer).
2. All subsequences have a length of 3 or **more**.

Return true if you can split nums according to the above conditions, or false
otherwise.

A **subsequence** of an array is a new array that is formed from the original
array by deleting some (can be none) of the elements without disturbing the 
relative positions of the remaining elements. (i.e., [1,3,5] is a subsequence of
 [1,2,3,4,5] while [1,3,2] is not).

<img src="images/659_1.png" alt="Left" width="450"/>
<img src="images/659_2.png" alt="right" width="450"/>
<img src="images/659_3.png" alt="right" width="450"/>
<img src="images/659_4.png" alt="right" width="450"/>
<img src="images/659_5.png" alt="right" width="450"/>

In [12]:
def isPossible(self, nums):
    """
    Determines if it's possible to split the sorted integer array 'nums' into one or more subsequences
    such that each subsequence is a consecutive increasing sequence of length at least 3.

    Parameters:
    - nums: List[int] - nums sorted list of integers.

    Returns:
    - bool: True if possible to split as per the conditions, False otherwise.
    """
    import collections  # Import the 'collections' module to use 'Counter'.

    # 'frequency_map' counts the number of times each number appears in 'nums'.
    # It represents the numbers that are still available to be placed in subsequences.
    frequency_map = collections.Counter(nums)

    # 'end' counts the number of subsequences ending with a specific number.
    # It helps track subsequences that can be extended.
    subsequence_map = collections.Counter()

    # Iterate over each number 'i' in the array 'nums'.
    for i in nums:
        # If 'i' has already been used up in subsequences, skip it.
        if not frequency_map[i]:
            continue
        
        frequency_map[i] -= 1    # Use one occurrence of 'i'.

        # Check if there's a subsequence ending with 'i - 1' that can be extended.
        if subsequence_map[i - 1] > 0:
            subsequence_map[i - 1] -= 1 # Decrement the count of subsequences ending with 'i - 1'.
            subsequence_map[i] += 1      # Increment the count of subsequences now ending with 'i'.

        # If no such subsequence exists, try to create a new subsequence [i, i+1, i+2].
        elif frequency_map[i + 1] > 0 and frequency_map[i + 2] > 0:
            frequency_map[i + 1] -= 1 # Use one occurrence of 'i + 1' 
            frequency_map[i + 2] -= 1 # Use one occurrence of 'i + 2'
            subsequence_map[i + 2] += 1 # Increment the count of subsequences ending with 'i + 2'.

        else:
            # Cannot place 'i' in any subsequence as per the rules.
            return False  # Early exit since the conditions cannot be met.

    # All numbers have been placed successfully in valid subsequences.
    return True

In [14]:
import collections
nums = [1,2,3,3,4,4,5,5]
frequency_map = collections.Counter(nums)
frequency_map

Counter({3: 2, 4: 2, 5: 2, 1: 1, 2: 1})

In [17]:
i = 1

frequency_map[i] -= 1




In [19]:
frequency_map[i + 1] -= 1 # Use one occurrence of 'i + 1' 
frequency_map[i + 2] -= 1 # Use one occurrence of 'i + 2'
subsequence_map[i + 2] += 1 # Increment the count of subsequences ending with 'i + 2'.

In [20]:
frequency_map

Counter({4: 2, 5: 2, 3: 1, 1: 0, 2: 0})

In [21]:
subsequence_map

Counter({3: 1})